Replace IncomingTransactionsController with helper

matthewwalsh0 opened the pull request:
## Explanation

The core `TransactionController` does not rely on a separate `IncomingTransactionsController` but instead internally fetches and processes incoming transaction data from Etherscan. As a result, incoming transactions in the core `TransactionController` reside in the same central `transactions` state array.

In order to align the extension controller with the core `TransactionController` as part of the controller migration, this PR:

- Removes the `IncomingTransactionController` and replaces it with an internal helper class.
- Moves incoming transactions into the central `transactions` object.
- Creates a new `RemoteTransactionSource` interface to decouple incoming transaction support from Etherscan.
- Splits the incoming transaction logic into multiple files for easier maintenance:
  - `etherscan.ts` - To retrieve data from Etherscan.
  - `EtherscanRemoteTransactionSource` - To fetch and normalise the Etherscan data.
  - `IncomingTransactionHelper` - To manage the block tracker listener and orchestrate the process.

Note that these same files and their logic (as unchanged as possible) already exist in the core repository.

Closes [#966](https://github.com/MetaMask/MetaMask-planning/issues/966)

## Manual Testing Steps

All existing incoming transaction support should be unchanged.

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

matthewwalsh0 committed:
Replace incoming transaction controller with helper

matthewwalsh0 committed:
Add unit tests for IncomingTransactionHelper

matthewwalsh0 committed:
Fix linting

matthewwalsh0 committed:
Store incoming transactions in transactions object

matthewwalsh0 committed:
Fix e2e test

matthewwalsh0 committed:
Fix unit tests

brad-decker commented:
@DDDDDanica @matthewwalsh0 just as a heads up @DDDDDanica is currently working on per-network preferences for incoming transactions and there may be some merge conflicts between your work. No action required right now just informing

matthewwalsh0 committed:
Create migration

Add related unit tests.

matthewwalsh0 committed:
Fix unit test

matthewwalsh0 committed:
Merge branch 'develop' into refactor/966-incoming-transaction-helper

matthewwalsh0 committed:
Use constant to set domain and subdomain

matthewwalsh0 committed:
Fix linting

matthewwalsh0 committed:
Fix linting

matthewwalsh0 committed:
Fix unit tests

matthewwalsh0 committed:
Refactor helper

Support API key in Etherscan remote transaction source.
Add unit tests for controller listeners.

matthewwalsh0 committed:
Update Lavamoat policies

matthewwalsh0 committed:
Revert "Update Lavamoat policies"

This reverts commit f087e632341c486d0c6d447ddab4e97159973328.

matthewwalsh0 committed:
Remove mutex

OGPoyraz commented on app/scripts/controllers/transactions/IncomingTransactionHelper.ts:
```diff
@@ -0,0 +1,271 @@
+
+      this.hub.emit('transactions', {
+        added: newTransactions,
+        updated: updatedTransactions,
```
Do we use "updated" txs somewhere or I missed it?

OGPoyraz commented on app/scripts/controllers/transactions/IncomingTransactionHelper.ts:
```diff
@@ -0,0 +1,271 @@
+    );
+
+    if (newTransactions.length > 0 || updatedTransactions.length > 0) {
+      this.#sortTransactionsByTime(newTransactions);
```
I know that this sort comes from the old code. But I couldn't get the idea of sorting txs since they aggregated into map on the listener, maybe you have an idea here?

matthewwalsh0 replied:
You're correct it's less useful in the extension given the map of transactions in state, but still good practice to provide the arrays in the event in a consistent order.

OGPoyraz commented on app/scripts/controllers/transactions/IncomingTransactionHelper.ts:
```diff
@@ -0,0 +1,271 @@
+  }
+
+  async #update(latestBlockNumberHex?: Hex): Promise<void> {
+    if (!this.#canStart()) {
```
Is there any situation that requires this check to be in "#update" since it's been checked on "start"? Or just an extra security?

matthewwalsh0 replied:
Once the polling is started via the `start` method, this method will be ran each time there is a new block so we still have to verify that incoming transactions are still enabled and that the current network is still supported in case they have been changed in the meantime.

The polling is also stopped and started when the preferences controller detects this has changed, but the controller (which will eventually be in core) can't assume the client is checking that.

matthewwalsh0 committed:
Merge branch 'develop' into refactor/966-incoming-transaction-helper

OGPoyraz commented on app/scripts/controllers/incoming-transactions.js:
```diff
@@ -1,320 +0,0 @@
-        if (currSelectedAddress === prevSelectedAddress) {
-          return;
-        }
-        await this._update(currSelectedAddress);
```
I liked the idea of removing all these subscription methods of controllers (`preferencesController`, `onboardingController`) and using callbacks instead to read the state. But on these subscription methods, we enforce `update`s. Did we do this intentionally or a minor thing that we can avoid?

matthewwalsh0 replied:
If you mean why do we no longer do this, it's just a matter of simplifying the code.

We previously used to listen to every possible controller which could change the address and maintain a copy in our own state.

Instead we now just need a single callback to request the current address from the state whenever we need to check it.

matthewwalsh0 replied:
Great spot, have added these updates back (plus unit tests) but via listeners in `MetaMaskController` to decouple it from the helper as it's a client decision.

matthewwalsh0 committed:
Update incoming transactions on account or network change

matthewwalsh0 committed:
Merge branch 'develop' into refactor/966-incoming-transaction-helper

matthewwalsh0 committed:
Merge branch 'develop' into refactor/966-incoming-transaction-helper

matthewwalsh0 committed:
Merge branch 'develop' into refactor/966-incoming-transaction-helper

matthewwalsh0 committed:
Fix E2E tests

matthewwalsh0 committed:
Merge branch 'develop' into refactor/966-incoming-transaction-helper
